mitiq/shadows/quantum_processing.py

```diff
@@ -0,0 +1,140 @@
+
+
+def generate_random_pauli_strings(
+    num_qubits: int, num_strings: int
+) -> List[str]:
+    """Generate a list of random Pauli strings.
+
+    Args:
+        num_qubits: The number of qubits in the Pauli strings.
+        num_strings: The number of Pauli strings to generate.
+
+    Returns:
+        A list of random Pauli strings.
+    """
+
+    # Sample random Pauli operators uniformly from ("X", "Y", "Z")
+    unitary_ensemble = ["X", "Y", "Z"]
+    paulis = np.random.choice(unitary_ensemble, (num_strings, num_qubits))
+    return ["".join(pauli) for pauli in paulis]
+
+
+def get_rotated_circuits(
+    circuit: cirq.Circuit,
+    pauli_strings: List[str],
+    add_measurements: bool = True,
+) -> List[cirq.Circuit]:
+    """Returns a list of circuits that are identical to the input circuit,
+    except that each one has single-qubit Clifford gates followed by
+    measurement gates that are designed to measure the input
+    Pauli strings in the Z basis.
+
+    Args:
+        circuit: The circuit to measure.
+        pauli_strings: The Pauli strings to measure in each output circuit.
+        add_measurements: Whether to add measurement gates to the circuit.
+
+    Returns:
+         The list of circuits with rotation and measurement gates appended.
+    """
+    qubits = sorted(list(circuit.all_qubits()))
+    rotated_circuits = []
+    for pauli_string in pauli_strings:
+        rotated_circuit = circuit.copy()
+        for qubit, pauli in zip(qubits, pauli_string):
+            # Pauli X measurement is equivalent to H plus a Z measurement
+            if pauli == "X":
+                rotated_circuit.append(cirq.H(qubit))
+            # Pauli X measurement is equivalent to S^-1*H plus a Z measurement
+            elif pauli == "Y":
+                rotated_circuit.append(cirq.S(qubit) ** -1)
+                rotated_circuit.append(cirq.H(qubit))
+            # Pauli Z measurement
+            else:
+                assert (
+                    pauli == "Z"
+                ), f"Pauli must be X, Y, Z. Got {pauli} instead."
+        if add_measurements:
+            rotated_circuit.append(cirq.measure(*qubits))
+        rotated_circuits.append(rotated_circuit)
+    return rotated_circuits
+
+
+def random_pauli_measurement(
+    circuit: cirq.Circuit,
+    n_total_measurements: int,
+    executor: Callable[[cirq.Circuit], MeasurementResult],
+) -> Tuple[NDArray[Any], NDArray[Any]]:
+    r"""
+    Given a circuit, perform random Pauli measurements on the circuit and
+    return outcomes. The outcomes are represented as a string where a
+    z-basis measurement outcome of :math:`|0\rangle` corresponds to 1, and
+    :math:`|1\rangle` corresponds to -1.
+
+    Args:
+        circuit: Cirq circuit.
+        n_total_measurements: number of snapshots.
+        executor: A callable which runs a circuit and returns a single
+            bitstring. If more bitstrings are returned, only one of them
+            is actually used.
```

Comment:
What do you think about adding a warning here in the API doc?
```suggestion
        executor: A callable which runs a circuit and returns a single
            bitstring.

    Warning:
        If your ``executor`` returns statistics for more than a single shot
        (i.e. returns counts for multiple bitstrings), a **single bitstring**
        will be chosen at random (from your distribution).
```